File: examples/callbacks/tests/bindings/test_callbacks.py
import unittest
import asyncio
from callbacks import *

class CallAnswererImpl(CallAnswerer):
    def __init__(self, mode):
        self.mode = mode

    def answer(self):
        if self.mode == "ready":
            return "Bonjour"
        elif self.mode == "busy":
            raise TelephoneError.Busy()
        else:
            raise ValueError("Testing an unexpected error")

class CallbacksTest(unittest.TestCase):
    def test_answer(self):
        cb_object = CallAnswererImpl("ready")
        telephone = Telephone()
        self.assertEqual("Bonjour", telephone.call(cb_object))

    def test_busy(self):
        cb_object = CallAnswererImpl("busy")
        telephone = Telephone()
        with self.assertRaises(TelephoneError.Busy):
            telephone.call(cb_object)

    def test_unexpected_error(self):
        cb_object = CallAnswererImpl("something-else")
        telephone = Telephone()
        with self.assertRaises(TelephoneError.InternalTelephoneError):
            telephone.call(cb_object)

class ProgressReporterImpl(ProgressReporter):
    def __init__(self, on_complete):
        self.on_complete = on_complete

    def report_progress(self, progress):
        print(f"Progress: {progress * 100}%")
        if progress == 1.0:
            self.on_complete()

class ProgressReporterTest(unittest.TestCase):
    def test_threaded_calls(self):
        def run_huge_task_wrapper():
            queue = asyncio.Queue()
            async def get():
                while True:
                    try:
                        queue.get_nowait()
                        return
                    except asyncio.QueueEmpty:
                        await asyncio.sleep(0.1)
            run_huge_task(ProgressReporterImpl(lambda: queue.put_nowait(None)))
            return get()
        asyncio.run(run_huge_task_wrapper())

unittest.main()

File: examples/callbacks/tests/bindings/test_callbacks.swift
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

import Dispatch

#if canImport(callbacks)
    import callbacks
#endif

struct SomeOtherError: Error { }

// Simple example just to see it work.
// Pass in a string, get a string back.
// Pass in nothing, get unit back.
class OnCallAnsweredImpl : CallAnswerer {
    var mode: String

    init(withMode: String) {
        mode = withMode
    }

    func answer() throws -> String {
        switch mode {
            case "ready":
                return "Bonjour"

            case "busy":
                throw TelephoneError.Busy(message: "I'm Busy")

            default:
                throw SomeOtherError()
        }
    }
}

let telephone = Telephone()

assert(try! telephone.call(answerer: OnCallAnsweredImpl(withMode: "ready")) == "Bonjour")

do {
    _ = try telephone.call(answerer: OnCallAnsweredImpl(withMode: "busy"))
} catch TelephoneError.Busy {
    // Expected error
}

do {
    _ = try telephone.call(answerer: OnCallAnsweredImpl(withMode: "unexpected-value"))
} catch TelephoneError.InternalTelephoneError {
    // Expected error
}

class ProgressReporterImpl : ProgressReporter {
    let continuation: CheckedContinuation<(), Never>
    init(withContinuation: CheckedContinuation<(), Never>) {
        continuation = withContinuation
    }
    func reportProgress(progress: Float) {
        print("Progress: \(progress * 100)%")
        if (progress == 1.0) {
            continuation.resume(returning: ())
        }
    }
}

let group = DispatchGroup()
group.enter()
Task {
    await withCheckedContinuation {
        runHugeTask(reporter: ProgressReporterImpl(withContinuation: $0))
    }
    group.leave()
}
group.wait()